### PR TITLE
Bluetooth: Mesh: Add support feature for proxy client

### PR DESCRIPTION
**Pull request by LingaoM**
1. Split proxy.c to proxy_common.c and proxy_server.c
2. Add Proxy Cient functionality to proxy_client.c

Add support for scan and parse network_id or node_id
   beacon, and connect to proxy server. discovery profile services.

Signed-off-by: Lingao Meng <mengabc1086@gmail.com>

### CODE REVIEW

**Review by trond-snekvik — CHANGES_REQUESTED** [review]
Great work!

I think the split to gatt_services.c and the beacon scanning are the only major problems here. 

The split to gatt_services isn't really relevant to the gatt client, and it kills git blame for the entire proxy subsystem. 
I'd rather we do it later, and split between proxy and pb gatt instead.

The beacon scanning is incredibly expensive, and I don't think it needs to be. IMO, the connect function should take mesh parameters, instead of forcing the application to implement the scan callback, as the Bluetooth address of the node it connects to is virtually never interesting. See comments for a detailed proposal.

File: include/bluetooth/mesh/proxy.h [hunk 1]
```diff
@@ -31,6 +31,73 @@ extern "C" {
  */
 int bt_mesh_proxy_identity_enable(void);
```
[the remark this answers is not in the document]
Comment (LingaoM) [reply include/bluetooth/mesh/proxy.h:32]: According Mesh Spec 1.0.1, even Proxy feature disabled, the node id also can be enabled. So, i'm think change to 
"This API requires that GATT Proxy Server supported".

File: subsys/bluetooth/mesh/proxy_common.c [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+		bt_mesh_pb_gatt_recv(object->conn, &object->buf);
+		break;
+#endif
+	default:
+		BT_WARN("Unhandled Message Type 0x%02x", object->msg_type);
+		break;
+	}
+
+	net_buf_simple_reset(&object->buf);
+}
```
[the remark this answers is not in the document]
Comment (LingaoM) [reply subsys/bluetooth/mesh/proxy_common.c:136]: I don't see the callback?

[the remark this answers is not in the document]
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/proxy_common.c:136]: I mean in the proxy_cfg function, which uses callbacks in the `object` to forward information. Doing this in only for some of the function calls kind of defeats the purpose

**Comment by LingaoM** [comment]
Mark as Draft, after solving the above problems, I hope to clarify the logical relationship.

**Comment by carlescufi** [comment]
@LingaoM will you pick this up again?

**Comment by LingaoM** [comment]
> @LingaoM will you pick this up again?

Yes, I didn't forget, I just waited for the free time.

**Comment by carlescufi** [comment]
@LingaoM please reopen this whenever you are ready to continue work on it.

**Comment by LingaoM** [comment]
> @LingaoM please reopen this whenever you are ready to continue work on it.

Sure.

**Comment by LingaoM** [comment]
> @LingaoM please reopen this whenever you are ready to continue work on it.

Please move to this:https://github.com/zephyrproject-rtos/zephyr/pull/41427/

**Pull request closed without merging**